perl/5.26.0/share/man/man1/perl.1
.\" Automatically generated by Pod::Man 4.09 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.if !\nF .nr F 0
.if \nF>0 \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    if !\nF==2 \{\
.        nr % 0
.        nr F 2
.    \}
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "PERL 1"
.TH PERL 1 "2017-05-30" "perl v5.26.0" "Perl Programmers Reference Guide"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
perl \- The Perl 5 language interpreter
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBperl\fR	[\ \fB\-sTtuUWX\fR\ ]
	[\ \fB\-hv\fR\ ]\ [\ \fB\-V\fR[:\fIconfigvar\fR]\ ]
	[\ \fB\-cw\fR\ ]\ [\ \fB\-d\fR[\fBt\fR][:\fIdebugger\fR]\ ]\ [\ \fB\-D\fR[\fInumber/list\fR]\ ]
	[\ \fB\-pna\fR\ ]\ [\ \fB\-F\fR\fIpattern\fR\ ]\ [\ \fB\-l\fR[\fIoctal\fR]\ ]\ [\ \fB\-0\fR[\fIoctal/hexadecimal\fR]\ ]
	[\ \fB\-I\fR\fIdir\fR\ ]\ [\ \fB\-m\fR[\fB\-\fR]\fImodule\fR\ ]\ [\ \fB\-M\fR[\fB\-\fR]\fI'module...'\fR\ ]\ [\ \fB\-f\fR\ ]
	[\ \fB\-C\ [\f(BInumber/list\fB]\ \fR]
	[\ \fB\-S\fR\ ]
	[\ \fB\-x\fR[\fIdir\fR]\ ]
	[\ \fB\-i\fR[\fIextension\fR]\ ]
	[\ [\fB\-e\fR|\fB\-E\fR]\ \fI'command'\fR\ ]\ [\ \fB\-\-\fR\ ]\ [\ \fIprogramfile\fR\ ]\ [\ \fIargument\fR\ ]...
.PP
For more information on these options, you can run \f(CW\*(C`perldoc perlrun\*(C'\fR.
.SH "GETTING HELP"
.IX Header "GETTING HELP"
The \fIperldoc\fR program gives you access to all the documentation that comes
with Perl.  You can get more documentation, tutorials and community support
online at <http://www.perl.org/>.
.PP
If you're new to Perl, you should start by running \f(CW\*(C`perldoc perlintro\*(C'\fR,
which is a general intro for beginners and provides some background to help
you navigate the rest of Perl's extensive documentation.  Run \f(CW\*(C`perldoc
perldoc\*(C'\fR to learn more things you can do with \fIperldoc\fR.
.PP
For ease of access, the Perl manual has been split up into several sections.
.SS "Overview"
.IX Subsection "Overview"
.Vb 4
\&    perl                Perl overview (this section)
\&    perlintro           Perl introduction for beginners
\&    perlrun             Perl execution and options
\&    perltoc             Perl documentation table of contents
.Ve
.SS "Tutorials"
.IX Subsection "Tutorials"
.Vb 3
\&    perlreftut          Perl references short introduction
\&    perldsc             Perl data structures intro
\&    perllol             Perl data structures: arrays of arrays
\&
\&    perlrequick         Perl regular expressions quick start
\&    perlretut           Perl regular expressions tutorial
\&
\&    perlootut           Perl OO tutorial for beginners
\&
\&    perlperf            Perl Performance and Optimization Techniques
\&
\&    perlstyle           Perl style guide
\&
\&    perlcheat           Perl cheat sheet
\&    perltrap            Perl traps for the unwary
\&    perldebtut          Perl debugging tutorial
\&
\&    perlfaq             Perl frequently asked questions
\&      perlfaq1          General Questions About Perl
\&      perlfaq2          Obtaining and Learning about Perl
\&      perlfaq3          Programming Tools
\&      perlfaq4          Data Manipulation
\&      perlfaq5          Files and Formats
\&      perlfaq6          Regexes
\&      perlfaq7          Perl Language Issues
\&      perlfaq8          System Interaction
\&      perlfaq9          Networking
.Ve
.SS "Reference Manual"
.IX Subsection "Reference Manual"
.Vb 10
\&    perlsyn             Perl syntax
\&    perldata            Perl data structures
\&    perlop              Perl operators and precedence
\&    perlsub             Perl subroutines
\&    perlfunc            Perl built\-in functions
\&      perlopentut       Perl open() tutorial
\&      perlpacktut       Perl pack() and unpack() tutorial
\&    perlpod             Perl plain old documentation
\&    perlpodspec         Perl plain old documentation format specification
\&    perlpodstyle        Perl POD style guide
\&    perldiag            Perl diagnostic messages
\&    perldeprecation     Perl deprecations
\&    perllexwarn         Perl warnings and their control
\&    perldebug           Perl debugging
\&    perlvar             Perl predefined variables
\&    perlre              Perl regular expressions, the rest of the story
\&    perlrebackslash     Perl regular expression backslash sequences
\&    perlrecharclass     Perl regular expression character classes
\&    perlreref           Perl regular expressions quick reference
\&    perlref             Perl references, the rest of the story
\&    perlform            Perl formats
\&    perlobj             Perl objects
\&    perltie             Perl objects hidden behind simple variables
\&      perldbmfilter     Perl DBM filters
\&
\&    perlipc             Perl interprocess communication
\&    perlfork            Perl fork() information
\&    perlnumber          Perl number semantics
\&
\&    perlthrtut          Perl threads tutorial
\&
\&    perlport            Perl portability guide
\&    perllocale          Perl locale support
\&    perluniintro        Perl Unicode introduction
\&    perlunicode         Perl Unicode support
\&    perlunicook         Perl Unicode cookbook
\&    perlunifaq          Perl Unicode FAQ
\&    perluniprops        Index of Unicode properties in Perl
\&    perlunitut          Perl Unicode tutorial
\&    perlebcdic          Considerations for running Perl on EBCDIC platforms
\&
\&    perlsec             Perl security
\&
\&    perlmod             Perl modules: how they work
\&    perlmodlib          Perl modules: how to write and use
\&    perlmodstyle        Perl modules: how to write modules with style
\&    perlmodinstall      Perl modules: how to install from CPAN
\&    perlnewmod          Perl modules: preparing a new module for distribution
\&    perlpragma          Perl modules: writing a user pragma
\&
\&    perlutil            utilities packaged with the Perl distribution
\&
\&    perlfilter          Perl source filters
\&
\&    perldtrace          Perl\*(Aqs support for DTrace
\&
\&    perlglossary        Perl Glossary
.Ve
.SS "Internals and C Language Interface"
.IX Subsection "Internals and C Language Interface"
.Vb 11
\&    perlembed           Perl ways to embed perl in your C or C++ application
\&    perldebguts         Perl debugging guts and tips
\&    perlxstut           Perl XS tutorial
\&    perlxs              Perl XS application programming interface
\&    perlxstypemap       Perl XS C/Perl type conversion tools
\&    perlclib            Internal replacements for standard C library functions
\&    perlguts            Perl internal functions for those doing extensions
\&    perlcall            Perl calling conventions from C
\&    perlmroapi          Perl method resolution plugin interface
\&    perlreapi           Perl regular expression plugin interface
\&    perlreguts          Perl regular expression engine internals
\&
\&    perlapi             Perl API listing (autogenerated)
\&    perlintern          Perl internal functions (autogenerated)
\&    perliol             C API for Perl\*(Aqs implementation of IO in Layers
\&    perlapio            Perl internal IO abstraction interface
\&
\&    perlhack            Perl hackers guide
\&    perlsource          Guide to the Perl source tree
\&    perlinterp          Overview of the Perl interpreter source and how it works
\&    perlhacktut         Walk through the creation of a simple C code patch
\&    perlhacktips        Tips for Perl core C code hacking
\&    perlpolicy          Perl development policies
\&    perlgit             Using git with the Perl repository
.Ve
.SS "Miscellaneous"
.IX Subsection "Miscellaneous"
.Vb 2
\&    perlbook            Perl book information
\&    perlcommunity       Perl community information
\&
\&    perldoc             Look up Perl documentation in Pod format
\&
\&    perlhist            Perl history records
\&    perldelta           Perl changes since previous version
\&    perl5241delta       Perl changes in version 5.24.1
\&    perl5240delta       Perl changes in version 5.24.0
\&    perl5223delta       Perl changes in version 5.22.3
\&    perl5222delta       Perl changes in version 5.22.2
\&    perl5221delta       Perl changes in version 5.22.1
\&    perl5220delta       Perl changes in version 5.22.0
\&    perl5203delta       Perl changes in version 5.20.3
\&    perl5202delta       Perl changes in version 5.20.2
\&    perl5201delta       Perl changes in version 5.20.1
\&    perl5200delta       Perl changes in version 5.20.0
\&    perl5184delta       Perl changes in version 5.18.4
\&    perl5182delta       Perl changes in version 5.18.2
\&    perl5181delta       Perl changes in version 5.18.1
\&    perl5180delta       Perl changes in version 5.18.0
\&    perl5163delta       Perl changes in version 5.16.3
\&    perl5162delta       Perl changes in version 5.16.2
\&    perl5161delta       Perl changes in version 5.16.1
\&    perl5160delta       Perl changes in version 5.16.0
\&    perl5144delta       Perl changes in version 5.14.4
\&    perl5143delta       Perl changes in version 5.14.3
\&    perl5142delta       Perl changes in version 5.14.2
\&    perl5141delta       Perl changes in version 5.14.1
\&    perl5140delta       Perl changes in version 5.14.0
\&    perl5125delta       Perl changes in version 5.12.5
\&    perl5124delta       Perl changes in version 5.12.4
\&    perl5123delta       Perl changes in version 5.12.3
\&    perl5122delta       Perl changes in version 5.12.2
\&    perl5121delta       Perl changes in version 5.12.1
\&    perl5120delta       Perl changes in version 5.12.0
\&    perl5101delta       Perl changes in version 5.10.1
\&    perl5100delta       Perl changes in version 5.10.0
\&    perl589delta        Perl changes in version 5.8.9
\&    perl588delta        Perl changes in version 5.8.8
\&    perl587delta        Perl changes in version 5.8.7
\&    perl586delta        Perl changes in version 5.8.6
\&    perl585delta        Perl changes in version 5.8.5
\&    perl584delta        Perl changes in version 5.8.4
\&    perl583delta        Perl changes in version 5.8.3
\&    perl582delta        Perl changes in version 5.8.2
\&    perl581delta        Perl changes in version 5.8.1
\&    perl58delta         Perl changes in version 5.8.0
\&    perl561delta        Perl changes in version 5.6.1
\&    perl56delta         Perl changes in version 5.6
\&    perl5005delta       Perl changes in version 5.005
\&    perl5004delta       Perl changes in version 5.004
\&
\&    perlexperiment      A listing of experimental features in Perl
\&
\&    perlartistic        Perl Artistic License
\&    perlgpl             GNU General Public License
.Ve
.SS "Language-Specific"
.IX Subsection "Language-Specific"
.Vb 4
\&    perlcn              Perl for Simplified Chinese (in EUC\-CN)
\&    perljp              Perl for Japanese (in EUC\-JP)
\&    perlko              Perl for Korean (in EUC\-KR)
\&    perltw              Perl for Traditional Chinese (in Big5)
.Ve
.SS "Platform-Specific"
.IX Subsection "Platform-Specific"
.Vb 10
\&    perlaix             Perl notes for AIX
\&    perlamiga           Perl notes for AmigaOS
\&    perlandroid         Perl notes for Android
\&    perlbs2000          Perl notes for POSIX\-BC BS2000
\&    perlce              Perl notes for WinCE
\&    perlcygwin          Perl notes for Cygwin
\&    perldos             Perl notes for DOS
\&    perlfreebsd         Perl notes for FreeBSD
\&    perlhaiku           Perl notes for Haiku
\&    perlhpux            Perl notes for HP\-UX
\&    perlhurd            Perl notes for Hurd
\&    perlirix            Perl notes for Irix
\&    perllinux           Perl notes for Linux
\&    perlmacos           Perl notes for Mac OS (Classic)
\&    perlmacosx          Perl notes for Mac OS X
\&    perlnetware         Perl notes for NetWare
\&    perlopenbsd         Perl notes for OpenBSD
\&    perlos2             Perl notes for OS/2
\&    perlos390           Perl notes for OS/390
\&    perlos400           Perl notes for OS/400
\&    perlplan9           Perl notes for Plan 9
\&    perlqnx             Perl notes for QNX
\&    perlriscos          Perl notes for RISC OS
\&    perlsolaris         Perl notes for Solaris
\&    perlsymbian         Perl notes for Symbian
\&    perlsynology        Perl notes for Synology
\&    perltru64           Perl notes for Tru64
\&    perlvms             Perl notes for VMS
\&    perlvos             Perl notes for Stratus VOS
\&    perlwin32           Perl notes for Windows
.Ve
.SS "Stubs for Deleted Documents"
.IX Subsection "Stubs for Deleted Documents"
.Vb 6
\&    perlboot            
\&    perlbot             
\&    perlrepository
\&    perltodo
\&    perltooc            
\&    perltoot
.Ve
.PP
On a Unix-like system, these documentation files will usually also be
available as manpages for use with the \fIman\fR program.
.PP
Some documentation is not available as man pages, so if a
cross-reference is not found by man, try it with perldoc.  Perldoc can
also take you directly to documentation for functions (with the \fB\-f\fR
switch). See \f(CW\*(C`perldoc \-\-help\*(C'\fR (or \f(CW\*(C`perldoc perldoc\*(C'\fR or \f(CW\*(C`man perldoc\*(C'\fR)
for other helpful options perldoc has to offer.
.PP
In general, if something strange has gone wrong with your program and you're
not sure where you should look for help, try making your code comply with
\&\fBuse strict\fR and \fBuse warnings\fR.  These will often point out exactly
where the trouble is.
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Perl officially stands for Practical Extraction and Report Language,
except when it doesn't.
.PP
Perl was originally a language optimized for scanning arbitrary
text files, extracting information from those text files, and printing
reports based on that information.  It quickly became a good language
for many system management tasks. Over the years, Perl has grown into
a general-purpose programming language. It's widely used for everything
from quick \*(L"one-liners\*(R" to full-scale application development.
.PP
The language is intended to be practical (easy to use, efficient,
complete) rather than beautiful (tiny, elegant, minimal).  It combines
(in the author's opinion, anyway) some of the best features of \fBsed\fR,
\&\fBawk\fR, and \fBsh\fR, making it familiar and easy to use for Unix users to
whip up quick solutions to annoying problems.  Its general-purpose
programming facilities support procedural, functional, and
object-oriented programming paradigms, making Perl a comfortable
language for the long haul on major projects, whatever your bent.
.PP
Perl's roots in text processing haven't been forgotten over the years.
It still boasts some of the most powerful regular expressions to be
found anywhere, and its support for Unicode text is world-class.  It
handles all kinds of structured text, too, through an extensive
collection of extensions.  Those libraries, collected in the \s-1CPAN,\s0
provide ready-made solutions to an astounding array of problems.  When
they haven't set the standard themselves, they steal from the best
\&\*(-- just like Perl itself.
.SH "AVAILABILITY"
.IX Header "AVAILABILITY"
Perl is available for most operating systems, including virtually
all Unix-like platforms.  See \*(L"Supported Platforms\*(R" in perlport
for a listing.
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
See perlrun.
.SH "AUTHOR"
.IX Header "AUTHOR"
Larry Wall <larry@wall.org>, with the help of oodles of other folks.
.PP
If your Perl success stories and testimonials may be of help to others 
who wish to advocate the use of Perl in their applications, 
or if you wish to simply express your gratitude to Larry and the 
Perl developers, please write to perl\-thanks@perl.org .
.SH "FILES"
.IX Header "FILES"
.Vb 1
\& "@INC"                 locations of perl libraries
.Ve
.PP
\&\*(L"@INC\*(R" above is a reference to the built-in variable of the same name;
see perlvar for more information.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
.Vb 4
\& http://www.perl.org/       the Perl homepage
\& http://www.perl.com/       Perl articles (O\*(AqReilly)
\& http://www.cpan.org/       the Comprehensive Perl Archive
\& http://www.pm.org/         the Perl Mongers
.Ve
.SH "DIAGNOSTICS"
.IX Header "DIAGNOSTICS"
Using the \f(CW\*(C`use strict\*(C'\fR pragma ensures that all variables are properly
declared and prevents other misuses of legacy Perl features.
.PP
The \f(CW\*(C`use warnings\*(C'\fR pragma produces some lovely diagnostics. One can
also use the \fB\-w\fR flag, but its use is normally discouraged, because
it gets applied to all executed Perl code, including that not under
your control.
.PP
See perldiag for explanations of all Perl's diagnostics.  The \f(CW\*(C`use
diagnostics\*(C'\fR pragma automatically turns Perl's normally terse warnings
and errors into these longer forms.
.PP
Compilation errors will tell you the line number of the error, with an
indication of the next token or token type that was to be examined.
(In a script passed to Perl via \fB\-e\fR switches, each
\&\fB\-e\fR is counted as one line.)
.PP
Setuid scripts have additional constraints that can produce error
messages such as \*(L"Insecure dependency\*(R".  See perlsec.
.PP
Did we mention that you should definitely consider using the \fBuse warnings\fR
pragma?
.SH "BUGS"
.IX Header "BUGS"
The behavior implied by the \fBuse warnings\fR pragma is not mandatory.
.PP
Perl is at the mercy of your machine's definitions of various
operations such as type casting, \fIatof()\fR, and floating-point
output with \fIsprintf()\fR.
.PP
If your stdio requires a seek or eof between reads and writes on a
particular stream, so does Perl.  (This doesn't apply to \fIsysread()\fR
and \fIsyswrite()\fR.)
.PP
While none of the built-in data types have any arbitrary size limits
(apart from memory size), there are still a few arbitrary limits:  a
given variable name may not be longer than 251 characters.  Line numbers
displayed by diagnostics are internally stored as short integers,
so they are limited to a maximum of 65535 (higher numbers usually being
affected by wraparound).
.PP
You may mail your bug reports (be sure to include full configuration
information as output by the myconfig program in the perl source
tree, or by \f(CW\*(C`perl \-V\*(C'\fR) to perlbug@perl.org .  If you've succeeded
in compiling perl, the perlbug script in the \fIutils/\fR subdirectory
can be used to help mail in a bug report.
.PP
Perl actually stands for Pathologically Eclectic Rubbish Lister, but
don't tell anyone I said that.
.SH "NOTES"
.IX Header "NOTES"
The Perl motto is \*(L"There's more than one way to do it.\*(R"  Divining
how many more is left as an exercise to the reader.
.PP
The three principal virtues of a programmer are Laziness,
Impatience, and Hubris.  See the Camel Book for why.


perl/5.26.0/.brew/perl.rb
class Perl < Formula
  desc "Highly capable, feature-rich programming language"
  homepage "https://www.perl.org/"
  url "https://www.cpan.org/src/5.0/perl-5.26.0.tar.xz"
  sha256 "9bf2e3d0d72aad77865c3bdbc20d3b576d769c5c255c4ceb30fdb9335266bf55"
  head "https://perl5.git.perl.org/perl.git", :branch => "blead"

  option "with-dtrace", "Build with DTrace probes"
  option "without-test", "Skip running the build test suite"

  deprecated_option "with-tests" => "with-test"

  def install
    if MacOS.version == :el_capitan && MacOS::Xcode.installed? && MacOS::Xcode.version >= "8.0"
      %w[cpan/IPC-Cmd/lib/IPC/Cmd.pm dist/Time-HiRes/Changes
         dist/Time-HiRes/HiRes.pm dist/Time-HiRes/HiRes.xs
         dist/Time-HiRes/Makefile.PL dist/Time-HiRes/fallback/const-c.inc
         dist/Time-HiRes/t/clock.t pod/perl588delta.pod
         pod/perlperf.pod].each do |f|
        inreplace f do |s|
          s.gsub! "clock_gettime", "perl_clock_gettime"
          s.gsub! "clock_getres", "perl_clock_getres", false
        end
      end
    end

    args = %W[
      -des
      -Dprefix=#{prefix}
      -Dprivlib=#{lib}/perl5/#{version}
      -Dsitelib=#{lib}/perl5/site_perl/#{version}
      -Dotherlibdirs=#{HOMEBREW_PREFIX}/lib/perl5/site_perl/#{version}
      -Dperlpath=#{opt_bin}/perl
      -Dstartperl=#!#{opt_bin}/perl
      -Dman1dir=#{man1}
      -Dman3dir=#{man3}
      -Duseshrplib
      -Duselargefiles
      -Dusethreads
    ]

    args << "-Dusedtrace" if build.with? "dtrace"
    args << "-Dusedevel" if build.head?

    system "./Configure", *args
    system "make"

    # OS X El Capitan's SIP feature prevents DYLD_LIBRARY_PATH from being
    # passed to child processes, which causes the make test step to fail.
    # https://rt.perl.org/Ticket/Display.html?id=126706
    # https://github.com/Homebrew/legacy-homebrew/issues/41716
    if MacOS.version < :el_capitan
      system "make", "test" if build.with? "test"
    end

    system "make", "install"
  end

  def caveats; <<-EOS.undent
    By default non-brewed cpan modules are installed to the Cellar. If you wish
    for your modules to persist across updates we recommend using `local::lib`.

    You can set that up like this:
      PERL_MM_OPT="INSTALL_BASE=$HOME/perl5" cpan local::lib
      echo 'eval "$(perl -I$HOME/perl5/lib/perl5 -Mlocal::lib)"' >> #{shell_profile}
    EOS
  end

  test do
    (testpath/"test.pl").write "print 'Perl is not an acronym, but JAPH is a Perl acronym!';"
    system "#{bin}/perl", "test.pl"
  end
end
